app/assets/javascripts/admin/addon/components/theme-settings-editor.js

```diff
@@ -0,0 +1,190 @@
+import Component from "@ember/component";
+import { action, computed } from "@ember/object";
+import { tracked } from "@glimmer/tracking";
+import { ajax } from "discourse/lib/ajax";
+import { inject as service } from "@ember/service";
+import I18n from "I18n";
+
+export default class ThemeSettingsEditor extends Component {
+  @service dialog;
+
+  @tracked editedContent = null;
+  @tracked errors = [];
+  @tracked saving = false;
+
+  didInsertElement() {
+    this.saving = false;
+    this.errors = [];
+  }
+
+  @computed("editedContent", "saving")
+  get saveButtonDisabled() {
+    return !this.documentChanged || this.saving;
+  }
+
+  @computed("editedContent")
+  get documentChanged() {
+    try {
+      if (!this.editedContent) {
+        return false;
+      }
+      const editedContentString = JSON.stringify(
+        JSON.parse(this.editedContent)
+      );
+      const themeSettingsString = JSON.stringify(this.condensedThemeSettings());
+      const same = editedContentString.localeCompare(themeSettingsString);
+      return same !== 0;
+    } catch (e) {
+      return true;
+    }
+  }
+
+  _theme() {
+    return this.model ? this.model.model : null;
+  }
+
+  condensedThemeSettings() {
+    if (!this._theme()) {
+      return null;
+    }
+    return this._theme().settings.map((setting) => ({
+      setting: setting.setting,
+      value: setting.value,
+    }));
+  }
+
+  @computed("theme")
+  get editorContents() {
+    return JSON.stringify(this.condensedThemeSettings(), null, "\t");
+  }
+
+  set editorContents(value) {
+    this.errors = [];
+    this.editedContent = value;
+  }
+
+  validateSettingsKeys(settings) {
+    return settings.reduce((acc, setting) => {
+      if (!acc) {
+        return acc;
+      }
```

Comment:
This is confusing... can the accumulator be falsey?

Reply:
yes - the idea is that we start out with a presumption that the input is valid (acc=true) ... then we loop through each setting and if we find one that is not valid, the acc becomes false, there is no point in continuing, and we show an error. I'm adding comments to make this function clearer.